### PR TITLE
Loop through paginated response to check for cluster id

### LINKED ISSUE #29712: EMRHook.get_cluster_id_by_name() doesn't use pagination
### Apache Airflow version

2.5.1

### What happened

When using EMRHook.get_cluster_id_by_name or any any operator that depends on it (e.g. EMRAddStepsOperator), if the results of the ListClusters API call is paginated (e.g. if your account has more than 50 clusters in the current region), and the desired cluster is in the 2nd page of results, None will be returned instead of the cluster ID.

### What you think should happen instead

Boto's pagination API should be used and the cluster ID should be returned.

### How to reproduce

Use `EmrAddStepsOperator` with the `job_flow_name` parameter on an `aws_conn_id` with more than 50 EMR clusters in the current region.

### Operating System

Linux

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==7.2.1

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by nsAstro**
Loops through the paginated emr list_clusters endpoint and searches each set for the given cluster_id.

closes: #29712

### CODE REVIEW

**Comment by Taragolis** [comment]
> I took a stab at this but I am at a loss as to how to test without spinning up 51 EMR clusters myself. If anyone has a suggestion for how to whitebox, mock, or test this, happy to contribute.

You could mock this method and use [side_effect](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock)

File: airflow/providers/amazon/aws/hooks/emr.py
Comment (thirtyseven) [inline airflow/providers/amazon/aws/hooks/emr.py]: This logic isn't quite right, what if two clusters have the same name but are on different pages?

File: airflow/providers/amazon/aws/hooks/emr.py
Comment (thirtyseven) [inline airflow/providers/amazon/aws/hooks/emr.py]: boto has a builtin pagination API, it would be nicer to use it rather than reimplement the logic:
https://boto3.amazonaws.com/v1/documentation/api/latest/guide/paginators.html

File: airflow/providers/amazon/aws/hooks/emr.py
Comment (nsAstro) [reply airflow/providers/amazon/aws/hooks/emr.py]: Addressed this in latest commit. I missed that cluster_name was non-unique.

File: airflow/providers/amazon/aws/hooks/emr.py
Comment (nsAstro) [reply airflow/providers/amazon/aws/hooks/emr.py]: Yeah, I was optimizing for speed when I thought we were filtering on a unique ID. Just rewrote to use the above.

**Review by thirtyseven — APPROVED** [review]

**Comment by thirtyseven** [comment]
@nsAstro I am not a codeowner, you'll need one of @eladkal @o-nikolas to review.

File: tests/providers/amazon/aws/hooks/test_emr.py
Comment (o-nikolas) [inline tests/providers/amazon/aws/hooks/test_emr.py]: Did you use `i % 50` so that the 0 case would be in the list twice? If so it's a bit sneaky and non-obvious IMHO.

File: tests/providers/amazon/aws/hooks/test_emr.py
Comment (o-nikolas) [inline tests/providers/amazon/aws/hooks/test_emr.py:247]: Should you not add a test case which covers the issue this PR is solving? I.e. A success case for getting a cluster id for a cluster that would be on the second page.

File: tests/providers/amazon/aws/hooks/test_emr.py
Comment (nsAstro) [reply tests/providers/amazon/aws/hooks/test_emr.py]: Yes I did. Was just a fast way to test, but I can add a comment to make it clearer or just hardcode a more explicit duplicated cluster name.

File: tests/providers/amazon/aws/hooks/test_emr.py
Comment (nsAstro) [reply tests/providers/amazon/aws/hooks/test_emr.py:247]: I'm not sure if list-clusters API on EMR is ordered at all, so I don't know if there's a way to guarantee that a given test-case cluster shows up on the second page only. If there's a way to guarantee this, I can add that as a test. I'll poke through the AWS docs to see if they provide some sort of ordering on how they list clusters to see if there's a deterministic test case I can write.

Comment (nsAstro) [reply tests/providers/amazon/aws/hooks/test_emr.py:247]: Their API [docs](https://docs.aws.amazon.com/emr/latest/APIReference/API_ListClusters.html) says unordered list. I could change the internal logic of the function to order the cluster, but that feels silly to do just to make a test.

Comment (vincbeck) [reply tests/providers/amazon/aws/hooks/test_emr.py:247]: I dont think that's what @o-nikolas is saying. I think he wants to have a test to cover the use case **in case** the cluster you are looking for is in the second page. The implementation should not change, just adding a test that covers this use case would increase the testing coverage. And I agree with that

Comment (nsAstro) [reply tests/providers/amazon/aws/hooks/test_emr.py:247]: I totally understand, I just am not sure how to force a given clusterName to be on the second page. Can I assume that even if the listClusters API provides an unordered response, it will provide that same order in subsequent queries? Then I can do a regular listClusters and fetch a cluster on the second page. But if the ordering is not consistent, I don't really have a surefire way to guarantee that the clusterName is on the second page. I could increase the number of clusters (<1000) to make the probability very high.

I'm happy to add tests, even for testing duplicate clusterNames which wasn't part of this PR. I'm just not too familiar with the EMR API, so I'm looking for some guidance.

Comment (nsAstro) [reply tests/providers/amazon/aws/hooks/test_emr.py:247]: I just realized with the way the marker works, that listClusters will be consistent. I'll edit the test accordingly.

File: tests/providers/amazon/aws/hooks/test_emr.py
Comment (vincbeck) [inline tests/providers/amazon/aws/hooks/test_emr.py:213]: LGTM but I would move it to a separate test for clarity purposes. The shorter the tests are, the more likely people will read them, the better :)

Comment (o-nikolas) [reply tests/providers/amazon/aws/hooks/test_emr.py:213]: +1

File: airflow/providers/amazon/aws/hooks/emr.py
Comment (uranusjr) [inline airflow/providers/amazon/aws/hooks/emr.py]:
```suggestion
        matching_clusters = [
            cluster
            for page in response_iterator
            for cluster in page["Clusters"]
            if cluster["Name"] == emr_cluster_name
        ]
```

File: tests/providers/amazon/aws/hooks/test_emr.py
Comment (uranusjr) [inline tests/providers/amazon/aws/hooks/test_emr.py]:
```suggestion

        Create 50 clusters with the range 1-50 appended and create a duplicate of index 1
        to test logic.
```

**Comment by o-nikolas** [comment]
Hey @nsAstro! Any status update on this one?

Seems like just a few small outstanding feedbacks.

**Comment by o-nikolas** [comment]
It doesn't look like @nsAstro is enganged on this one any more. I'll push the remaining fixes if I get some time in the next few days.

**Comment by o-nikolas** [comment]
I updated the PR to address the remaining feedback

File: tests/providers/amazon/aws/hooks/test_emr.py
Comment (o-nikolas) [reply tests/providers/amazon/aws/hooks/test_emr.py]: Addressed

File: tests/providers/amazon/aws/hooks/test_emr.py
Comment (o-nikolas) [reply tests/providers/amazon/aws/hooks/test_emr.py:247]: Addressed

File: tests/providers/amazon/aws/hooks/test_emr.py
Comment (o-nikolas) [reply tests/providers/amazon/aws/hooks/test_emr.py:213]: Addressed

**Comment by potiuk** [comment]
unexpected ident does not signal anything good.

**Comment by o-nikolas** [comment]
> unexpected ident does not signal anything good.

Ooop! I updated @nsAstro's PR through the Github UI text editor (which has no bells and whistles for linting or catching those indents) so I missed that one! Hopefully that stops the build from timing out :grin:

**Review by vincbeck — APPROVED** [review]

**Comment by o-nikolas** [comment]
Looks like this is all passing and we have a couple of approvals. I'm not sure if it's proper for me to approve and merge this since the PR is mostly my code at this point, so CCing @potiuk and @eladkal to approve and merge (and/or advise that it's okay if I do).

**Review by eladkal — APPROVED** [review]

**Pull request merged**
